rfcs/20221116-enable-stochastic-rounding.md

```diff
@@ -0,0 +1,335 @@
+
+$$
+sr(x) = \begin{cases} \lfloor{x}\rfloor,& \text{with probability }
+p_1(x)=1-\frac{x-\lfloor{x}\rfloor}{\delta}\\
+\lfloor{x}\rfloor + \delta, &
+\text{with probability } p_2(x)=\frac{x-\lfloor{x}\rfloor}{\delta}\\ \end{cases}
+$$
+
+Where $\delta$ indicates the greatest representable floating/fixed-point
+number less than or equal to x.
+
+## User Benefit
+
+Users can be benefited from following aspects:
+
+-   Ready-to-use API for performing accurate stochastic rounding on many
+    different floating point formats.
+-   Performance speedups on Google TPUs.
+
+## Design Proposal
+
+During quantization, users replace high-precision floating-point numbers with
+low-precision fixed-point numbers or lower-bit floats to reduce the
+computational cost of neural network training or inference.
+
+For example, below is a code snippet that represents a part of process to
+quantize a tensor from floating-points to int8, which is then used for matrix
+multiplications:
+
+```
+inputs = tf.clip_by_value(inputs, -128.0, 127.0)
+inputs = tf.math.round(inputs) inputs = tf.cast(inputs, dtype=tf.int8)
+// 'weights' tensor has been converted to int8 format beforehand.
+output = tf.einsum(inputs, weights)
+```
+
+For quantizing bigger floats to smaller floats, users need to do casting like
+below:
+
+```
+inputs = tf.cast(inputs, dtype=tf.bfloat16)
+// 'weights' tensor has been converted to bfloat16 format beforehand.
+output = tf.einsum(inputs, weights)
+```
+
+There are two essentials to achieve stochastic rounding:
+
+-   A random value generated from uniform distributions to emulate
+    stochasticity.
+-   Perform deterministic rounding on the operand with the rounding direction
+    determined by the given random value.
+
+Although stochastic rounding is probabilistic, the probability can be explicitly
+controlled by choosing an appropriate random number generator. Users might want
+to reproduce the rounding results for debugging or behavioral enforcement
+purposes. This usually requires a reproducible pseudo random number generator,
+whose output is deterministic by a given state. In tensorflow, this generator is
+exposed via the tf.random.generator API. Since the purpose of quantized training
+is to reduce the bits used in calculations, changing the data type after
+rounding is a necessity. Given these considerations, an API is ultimately
+proposed like below:
+
+```
+tf.stochastic_cast(input, dtype, random_seed=None, random_algo=’Philox’)
+```
+
+<table>
+ <tr><td><strong>Args</strong></td><td><strong>Definition</strong></td></tr>                                      
+ <tr><td>input</td><td>The original tensor to be casted</td></tr>                              
+ <tr><td>dtype</td><td>Desired type after casting</td></tr>       
+ <tr><td>random_seed</td><td>Random seed if users need to control the randomness. If None, the system will generate a seed.</td></tr>                              
```

Comment:
If the user has a set a seed with [`tf.random.set_seed`](https://www.tensorflow.org/api_docs/python/tf/random/set_seed), I think we should use that if `random_seed` is None. Typically ops that take a `seed` argument defaulting to None, like [`tf.random.normal`](https://www.tensorflow.org/api_docs/python/tf/random/normal), use the global seed from `tf.random.set_seed` if no seed is passed to the function.

Not all random ops use the global seed. There are three classes of random ops:

* Legacy random ops like `tf.random.normal`. These ops takes a seed defaulting to None. `tf.stochastic_cast` also follows this pattern. Legacy random ops will generate deterministic random numbers if either the seed argument or the global seed is set.
* [`tf.random.Generator`][rand-gen]: This doesn't use the global seed, but instead is created with the `from_seed` method, which requires passing in a seed, or the `from_non_deterministic_state` method, which nondeterministically creates a seed
* Stateless random ops like [`tf.random.stateless_uniform`](https://www.tensorflow.org/api_docs/python/tf/random/stateless_uniform). These ops require the user to pass a seed, and deterministically return the same values each time given the same inputs and seed.

Since `tf.stochastic_cast` is most similar to the legacy random ops, it's best to use the global seed from `tf.random.set_seed` to be consistent with them. The user expects that if they call `tf.random.set_seed`, then non-stateless non-generator random ops will use the global seed if no seed is passed.

This is important for determinism, which is enabled by calling [`tf.config.experimental.enable_op_determinism`](https://www.tensorflow.org/api_docs/python/tf/config/experimental/enable_op_determinism) and setting the global seed. Once the user enables determinism, they expect all ops to be deterministic.

CC @wangpengmit

[rand-gen]: https://www.tensorflow.org/api_docs/python/tf/random/Generator

Reply:
The 3 classes Reed listed are correct. I view `tf.stochastic_cast` as a new RNG op. For a new RNG op, we only recommend implementing it as the 3rd class (i.e. totally stateless). After that, the 2nd class can be (optionally) easily implemented as a thin wrapped on top of the 3rd class ([example](https://github.com/tensorflow/tensorflow/blob/20b6531762ca1949d25ecf858a0bb3de906df896/tensorflow/python/ops/stateful_random_ops.py#L719)). The 1st class (potentially along with `tf.random.set_seed` and `tf.compat.v1.get_seed`) is going to be deprecated because those ops store the RNG state in the op kernel, which interacts badly with other TF facilities (e.g. `tf.function`).

A detail is that to implement the 3rd class, we currently recommend the new op taking in a `key` and a `counter` ([example](https://github.com/tensorflow/tensorflow/blob/20b6531762ca1949d25ecf858a0bb3de906df896/tensorflow/core/ops/stateless_random_ops_v2.cc#L90)), instead of a `seed` ([example](https://github.com/tensorflow/tensorflow/blob/20b6531762ca1949d25ecf858a0bb3de906df896/tensorflow/core/ops/stateless_random_ops.cc#L57)), to make supporting the 2nd class easier. The Python API can still take in a `seed` ([example](https://github.com/tensorflow/tensorflow/blob/20b6531762ca1949d25ecf858a0bb3de906df896/tensorflow/python/ops/stateless_random_ops.py#L503)).

Reply:
Thanks both! To summarize:
We endorse following tf.stateless_uniform and tf.random.Generator's user usage and implementation over legacy tf.random.normal's, as the latter is incorporated badly in the new TF stack.

A.I.s:
1) Random seed to be required in the API level 
2) Use key + counter instead of single seed in the kernel